Pump Version python to 3.12, 3.13. Allow pip install for python 3.12, 3.13

maycuatroi opened the pull request:
# Overview

Update file `pyproject.toml` to enable pip install for python 3.12

# References

Example workbook work perfect when I pip install on python 3.12
![image](https://github.com/user-attachments/assets/42505c80-6629-427f-a9c9-023272a5c623)

maycuatroi commented:
It's strange, I didn't make any changes to the Python code, yet it's causing CI errors with mypy. 
@odashi-san, Do you think I should fix these CI errors?

maycuatroi commented:
@odashi-san. I fixed CI to the last version of mypy and black. Please approve the CI scan

odashi commented:
@maycuatroi Thanks!

I think you tweaked some code blocked by version switches, e.g.,

* if blocks conditioned by `if sys.version_info.minor < blah:`
* test blocks conditioned by `@test_utils.require_at_most(blah)`

These blocks are activated while the library is running on a conditioned version of Python, and the code block must follow the correct syntax on the corresponding version.

I guessed that this is a good time to drop support for legacy (EOL-ed) versions such as 3.7 and 3.8, by simply removing corresponding blocks from the library.
Sorry for bothering, but I'd quite appreciated it if you could take some additional efforts.

maycuatroi commented:
@odashi -san
I've checked and fixed the unit tests for the CI accordingly. Here's what I did:

- Removed support for Python 3.7 and 3.8 (since they've reached end-of-life).
- Tested with Python 3.13 – everything worked great!

I really enjoy using this library, and I appreciate your professionalism.

To ensure everything works across multiple Python versions, I also created a custom Docker Compose setup for running the CI tests on various versions 😄  (I'm not commit it yet):

```yaml
version: "3.8"

services:
  unit-tests:
    image: python:3.13
    environment:
      - PYTHONPATH=src
    volumes:
      - .:/app  # This will copy all files in the current directory into the /app directory in the container
    working_dir: /app
    command: >
      bash -c "
      python -m pip install --upgrade pip &&
      python -m pip install -e '.[dev]' &&
      python -m pytest src
      "

  black:
    image: python:3.13
    volumes:
      - .:/app  # Mounting the project directory
    working_dir: /app
    command: >
      bash -c "
      python -m pip install --upgrade pip &&
      python -m pip install black &&
      python -m black -v --check src
      "

  flake8:
    image: python:3.13
    volumes:
      - .:/app  # Mounting the project directory
    working_dir: /app
    command: >
      bash -c "
      python -m pip install --upgrade pip &&
      python -m pip install pyproject-flake8 &&
      pflake8 -v src
      "

  isort:
    image: python:3.13
    volumes:
      - .:/app  # Mounting the project directory
    working_dir: /app
    command: >
      bash -c "
      python -m pip install --upgrade pip &&
      python -m pip install isort &&
      python -m isort --check src
      "

  mypy:
    image: python:3.13
    volumes:
      - .:/app  # Mounting the project directory
    working_dir: /app
    command: >
      bash -c "
      python -m pip install --upgrade pip &&
      python -m pip install '.[mypy]' &&
      python -m mypy src
      "
```

odashi commented on src/latexify/test_utils.py:
```diff
@@ -93,8 +97,8 @@ def ast_equal(observed: ast.AST, expected: ast.AST) -> bool:
 
-    except (AssertionError, AttributeError):
-        return False
+    except (AssertionError, AttributeError) as e:
+        raise e  # raise to debug easier.
```
```suggestion
    except (AssertionError, AttributeError):
        raise  # raise to debug easier.
```

odashi commented on src/latexify/test_utils.py:
```diff
@@ -109,19 +113,7 @@ def assert_ast_equal(observed: ast.AST, expected: ast.AST) -> None:
-observed={ast.dump(observed)}
-expected={ast.dump(expected)}
-"""
+    indent = 4 if sys.version_info.minor >= 9 else None
+    assert ast_equal(
+        observed, expected
+    ), f"AST does not match. observed = {observed}\nexpected = {expected}"
```
I looks this change is incomplete and should be reverted for now

maycuatroi replied:
I removed the if condition because python always >=3.9 and revert the ast dump logic

odashi commented on src/latexify/ast_utils.py:
```diff
@@ -57,16 +57,10 @@ def make_constant(value: Any) -> ast.expr:
-        if isinstance(value, str):
-            return ast.Str(s=value)
-        if isinstance(value, bytes):
-            return ast.Bytes(s=value)
+        if value in {None, False, True, ...} or isinstance(
+            value, (int, float, complex, str, bytes)
+        ):
+            return ast.Constant(value=value)
     else:
```
As we decided to drop old versions, this if-block is no longer necessary.

odashi commented on src/latexify/ast_utils_test.py:
```diff
@@ -38,15 +38,15 @@ def test_make_attribute() -> None:
 @pytest.mark.parametrize(
```
Ditto: as this test case only runs on older versions than 3.8, we can simply remove it.

maycuatroi replied:
I’m not completely sure I understand this part. Could you help me clarify or commit to update? 
I noticed there were some test cases added earlier, but I’m unsure of their purpose, so I didn’t want to remove them just yet. I’ve updated the code to stop using the deprecated functions based on warnings from the latest version, but I wanted to double-check with you first.

maycuatroi replied:
I removed this test case

odashi commented on src/latexify/transformers/identifier_replacer.py:
```diff
@@ -55,21 +55,25 @@ def visit_FunctionDef(self, node: ast.FunctionDef) -> ast.FunctionDef:
                 kw_defaults=visited.args.kw_defaults,
                 defaults=visited.args.defaults,
-            )
+            )  # type: ignore
```
Could you resolve typing errors?
If it is difficult, could you use `#type: ignore[code]` rather than the wildcard `#type: ignore` to show which error code we ignored?

maycuatroi replied:
I removed the python 3.8 conditions

maycuatroi commented:
@odashi  -san
I Create new function `ast_function_def` to handle `type_params` for python 3.12 and I fixed the flake8 CI issue. Please help me review and approve workflow 🙏

maycuatroi commented:
@odashi -san. I think things look good 😄

odashi commented on src/latexify/ast_utils.py:
```diff
@@ -147,3 +147,9 @@ def extract_function_name_or_none(node: ast.Call) -> str | None:
     return None
+
+
+def ast_function_def(*args, **kwargs) -> ast.FunctionDef:
```
Please follow the convention of this library:

* function names should be basically starting with a verb
* add docstring (with Google's docstring style)
* add unit test for this function (in `ast_utils_test.py`).
* (if possible) full argument list is encouraged rather than `args` and `kwargs`.

maycuatroi replied:
Yes, perfect. I updated the coding convention and created the test case.
Kwargs current can't replace be cause `type_params` is not optional so we need pass through kwargs and filter in the wrapper function

odashi commented on src/latexify/ast_utils.py:
```diff
@@ -147,3 +147,9 @@ def extract_function_name_or_none(node: ast.Call) -> str | None:
+
+
+def ast_function_def(*args, **kwargs) -> ast.FunctionDef:
+    if sys.version_info.minor < 12:
+        kwargs.pop("type_params", None)
+    return ast.FunctionDef(*args, **kwargs)
```
I think we haven't use `type_params` anywhere with meaningful semantics, and this function can be defined to hide this variable from outer functions rather than passing it always.

maycuatroi replied:
We need `type_params` to pass the `mypy` scan

odashi replied:
How about:

```python
def this_fn(...):
    if sys.version_info.minor < 12:
        # We know that the following invocation is correct.
        # In this case the mypy error is just wrong and we need to ignore it.
        return ast.FunctionDef(...)  # type: ignore[some-appropriate-flag]
  
    return ast.FunctionDef(..., type_params=[])
```

maycuatroi replied:
I think mark ignore on this function is better. Let me update

maycuatroi replied:
I updated follow your comment

odashi commented on src/latexify/ast_utils_test.py:
```diff
@@ -83,7 +81,6 @@ def test_make_constant_invalid() -> None:
 
 
-@test_utils.require_at_most(7)
 @pytest.mark.parametrize(
```
ditto: this test case can be removed.

odashi commented on src/latexify/ast_utils_test.py:
```diff
@@ -114,7 +110,6 @@ def test_is_constant(value: ast.AST, expected: bool) -> None:
 
 
-@test_utils.require_at_most(7)
 @pytest.mark.parametrize(
```
ditto

odashi commented on src/latexify/ast_utils_test.py:
```diff
@@ -34,7 +34,6 @@ def test_make_attribute() -> None:
     )
 
 
-@test_utils.require_at_most(7)
```
This test case (the whole function) is no longer meaningful, and should be completely removed.

Same as every other test case with a decoration `@test_utils.require_at_most(7)`

maycuatroi replied:
I removed the legacy test case

odashi commented on src/latexify/parser_test.py:
```diff
@@ -7,6 +7,7 @@
 import pytest
 
 from latexify import exceptions, parser, test_utils
+from latexify.ast_utils import ast_function_def
```
We don't allow importing functions/classes directly, import module instead.

Same for every other code.

Reference: https://google.github.io/styleguide/pyguide.html#22-imports

```suggestion
from latexify import ast_utils
```

odashi reviewed (commented):
Cool, basically everything looks good. Comments are almost based on coding conventions.

maycuatroi commented:
@odashi  -san, please approve the workflow, I fixed the isort issue 😄

maycuatroi commented:
@odashi -san, please help me review. I updated follow your comment 😃

odashi commented on src/latexify/ast_utils_test.py:
```diff
@@ -83,25 +61,23 @@ def test_make_constant_invalid() -> None:
         (ast.Global(names=["qux"]), False),
     ],
 )
 def test_is_constant_legacy(value: ast.AST, expected: bool) -> None:
```
It looks there exists some test cases from that only `@test_utils.require_at_most(7)` was removed like this. Could you check the changes entirely and remove them?

maycuatroi replied:
The decorator (@test_utils.require_at_most(7)) was previously used to indicate that the test case would run on Python versions >= 3.7. So, even when we were on Python 3.9, the test case was still executed. Now that we've upgraded to at least Python 3.9, the test case should still run as intended. I believe it’s best to keep the test case as it is.

odashi replied:
`require_at_most(k)` requires versions equal or *less* than 3.k. the decorator does nothing otherwise.
https://github.com/google/latexify_py/blob/678cd0e463ee0cfd179fa0da2897e7e0535c3b46/src/latexify/test_utils.py#L51-L52

odashi commented on src/latexify/ast_utils.py:
```diff
@@ -173,3 +147,65 @@ def extract_function_name_or_none(node: ast.Call) -> str | None:
+
+
+def create_function_def(
+    name,
+    args,
+    body,
+    decorator_list,
+    returns=None,
+    type_comment=None,
+    type_params=None,
+    lineno=None,
+    col_offset=None,
+    end_lineno=None,
+    end_col_offset=None,
```
please add type hints.

odashi commented on src/latexify/ast_utils_test.py:
```diff
@@ -114,25 +90,23 @@ def test_is_constant(value: ast.AST, expected: bool) -> None:
         (ast.Global(names=["qux"]), False),
     ],
 )
 def test_is_str_legacy(value: ast.AST, expected: bool) -> None:
```
this test case as well

maycuatroi replied:
Same with decorator @test_utils.require_at_most(7)

odashi replied:
Yes, removing every test function with decoration `@test_utils.require_at_most(7)` should be fine.

maycuatroi replied:
@odashi  Yes, I removed all the test function decorated with @test_utils.require_at_most(7)

maycuatroi replied:
@odashi  please check and approve the Checks :D

odashi reviewed (commented):
It's almost done I think, but there are still some little comments (maybe the last ones)

odashi reviewed (approved):
LGTM
